File: 4283-multi-source-flood-fill/4283-multi-source-flood-fill.java
class Solution {
    private int time[][], vis[][], color[][];
    
    static class Tuple {
        int row, col, time;
        public Tuple(int row, int col, int time) {
            this.row = row;
            this.col = col;
            this.time = time;
        }
        @Override
        public String toString() {
            return "(" + row + " " + col + ")";
        }
    }

    static class customSort implements Comparator<Tuple> {
        @Override
        public int compare(Tuple first, Tuple second) {
            return Integer.compare(first.time, second.time);
        }
    }

    public int[][] colorGrid(int n, int m, int[][] sources) {
        time = new int[n][m];
        vis = new int[n][m];
        color = new int[n][m];

        for (int x[]: time)
            Arrays.fill(x, (int)(1e9));

        PriorityQueue<Tuple> pq = new PriorityQueue<>(new customSort());

        for (int curr[] : sources) {
            int currRow = curr[0], currCol = curr[1], currColor = curr[2];
            pq.offer(new Tuple(currRow, currCol, 0));
            vis[currRow][currCol] = 1;
            time[currRow][currCol] = 0;
            color[currRow][currCol] = currColor;
        }
        
        int dir[][] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        while (pq.size() > 0) {
            int currRow = pq.peek().row, currCol = pq.peek().col, currTime = pq.peek().time;
            pq.poll();
            for (int dire[]: dir) {
                int newRow = currRow + dire[0], newCol = currCol + dire[1];
                if (newRow < n && newCol < m && newRow >= 0 && newCol >= 0) {
                    if (vis[newRow][newCol] == 1) {
                        if (time[newRow][newCol] == currTime + 1) {
                            if (color[newRow][newCol] < color[currRow][currCol]) {
                                color[newRow][newCol] = color[currRow][currCol];
                                pq.offer(new Tuple(newRow, newCol, currTime + 1));
                            }
                        }
                    } else {
                        vis[newRow][newCol] = 1;
                        color[newRow][newCol] = color[currRow][currCol];
                        time[newRow][newCol] = currTime + 1;
                        pq.offer(new Tuple(newRow, newCol, currTime + 1)); 
                    
                    }
                }
            }
        }

        return color;
    }
}

File: 4283-multi-source-flood-fill/4283-multi-source-flood-fill.py
import heapq

class Tuple:
    def __init__(self, row, col, time):
        self.row = row
        self.col = col
        self.time = time
    
    def __repr__(self):
        return f"({self.row}, {self.col}, {self.time})"

    def __lt__(self, other):
        return self.time < other.time

class Solution:
    def colorGrid(self, n: int, m: int, sources: list[list[int]]) -> list[list[int]]:
        time = [[10 ** 9] * m for _ in range(n)]
        color = [[0] * m for _ in range(n)]
        vis = [[0] * m for _ in range(n)]

        pq = []
        for i in range(len(sources)):
            curr_row, curr_col, curr_color = sources[i][0], sources[i][1], sources[i][2]
            heapq.heappush(pq, Tuple(curr_row, curr_col, 0))
            color[curr_row][curr_col] = curr_color 
            time[curr_row][curr_col] = 0
            vis[curr_row][curr_col] = 1     

        dir = [[-1, 0], [1, 0], [0, -1], [0, 1]]        
        while pq:
            curr_tuple = heappop(pq)
            curr_row, curr_col, curr_time = curr_tuple.row, curr_tuple.col, curr_tuple.time
            for dire in dir:
                new_row, new_col, new_time = curr_row + dire[0], curr_col + dire[1], curr_time + 1
                if new_row >= 0 and new_row < n and new_col >= 0 and new_col < m:
                    if vis[new_row][new_col] == 1:
                        if time[new_row][new_col] == new_time:
                            if color[new_row][new_col] < color[curr_row][curr_col]:
                                color[new_row][new_col] = color[curr_row][curr_col]
                                heapq.heappush(pq, Tuple(new_row, new_col, new_time))
                    else:
                        vis[new_row][new_col] = 1
                        time[new_row][new_col] = curr_time + 1
                        color[new_row][new_col] = color[curr_row][curr_col]
                        heapq.heappush(pq, Tuple(new_row, new_col, time[new_row][new_col]))

        return color
